store/localstore/snapshot.go

```diff
@@ -48,65 +43,68 @@ func newSnapshot(store *dbStore, db engine.DB, ver kv.Version) *dbSnapshot {
 	return ss
 }
 
-func (s *dbSnapshot) internalSeek(startKey []byte) (engine.Iterator, error) {
+// mvccSeek seeks for the first key in db which has a k >= key and a version <=
+// snapshot's version. If strict is true, only k == key can be returned.
+// The returned slices should be cloned before modify.
+func (s *dbSnapshot) mvccSeek(key kv.Key, strict bool) (kv.Key, []byte, error) {
 	s.store.mu.RLock()
 	defer s.store.mu.RUnlock()
 
 	if s.store.closed {
-		return nil, errors.Trace(ErrDBClosed)
+		return nil, nil, errors.Trace(ErrDBClosed)
 	}
 
-	if s.rawIt == nil {
-		var err error
-		s.rawIt, err = s.db.Seek(minKey)
+	// Key layout:
+	// ...
+	// Key (Meta)      -- (1)
+	// Key_verMax      -- (2)
+	// ...
+	// Key_ver+1       -- (3)
+	// Key_ver         -- (4)
+	// Key_ver-1       -- (5)
+	// ...
+	// Key_0           -- (6)
+	// NextKey (Meta)  -- (7)
+	// ...
+	// EOF
+	for {
+		mvccKey := MvccEncodeVersionKey(key, s.version)
+		mvccK, v, err := s.db.Seek(mvccKey) // search for [4...EOF)
 		if err != nil {
-			return nil, errors.Trace(err)
+			if terror.ErrorEqual(err, engine.ErrNotFound) { // EOF
+				err = errors.Wrap(err, kv.ErrNotExist)
+			}
+			return nil, nil, errors.Trace(err)
 		}
+		k, _, err := MvccDecode(mvccK)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		if key.Cmp(k) != 0 { // currently on [7]
+			if strict {
+				return nil, nil, errors.Trace(kv.ErrNotExist)
+			}
+			// search for NextKey
+			key = k
+			continue
+		}
+		if isTombstone(v) { // current key is deleted
```

Comment:
If current key is deleted, we still need to seek for nextKey as the seek's definition.